### PR TITLE
Adding settings page with token storage table

### LINKED ISSUE #99: Token Storage: Settings page

### PR DESCRIPTION
**Pull request by riesentoaster**
## TODOs
- [x] Create storyboard elements
- [x] Move API access to backend of editor
- [x] Manual testing

## Related Issues

- Closes #99

## Definition of Done

- [x] User Story
  - [x] All Acceptance Criteria met
- [x] Code
  - [x] No more code needed
  - [x] No known bugs
- [x] Clean Code
  - [x] Comments have been created for new code
  - [x] ESLinter found no issues (editor & executor only)
  - [x] No unavoidable Code Smells
- [x] Testing
  - [x] Existing Tests pass
  - [x] New Unit Tests created
  - [x] New Unit Tests pass
- [x] SonarCloud Quality Gate passed

## Definition of more than Done (optional)

- [ ] Testing
  - [ ] New Integration Tests created
  - [ ] New Integration Tests passed

### CODE REVIEW

File: src/editor/components/flow/edit-dialog.tsx [hunk 1]
```diff
@@ -19,7 +19,7 @@ const states = [
 ]
 
 export const EditDataFlowDialog = ( { isEditOpen, setIsEditOpen, save, flow, setFlow, deleteFlow }: EditDataFlowDialogProps ): JSX.Element => {
-  const [deleteFlowName, setDeleteFlowName] = useState()
+  const [deleteFlowName, setDeleteFlowName] = useState( '' )
```
Comment (MrF3lix) [inline src/editor/components/flow/edit-dialog.tsx:22]: Why was this necessary? You didn't change anything else in this file?

File: src/editor/components/settings/token-insert-new.tsx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+            label="Token value"
+            value={value}
+            onChange={( e: React.ChangeEvent<HTMLInputElement> ): void => setValue( e.target.value )}
+          />
+        </div>
+        <div>
+          <Button
+            primary
+            disabled={!name || name.length === 0 || !value || value.length === 0}
+            onClick={(): void => saveNewToken( name, value )}
```
Comment (MrF3lix) [inline src/editor/components/settings/token-insert-new.tsx:35]: Reset the `name` and `value` if the token has been added sucessfully.

File: src/editor/components/settings/token-list-item.tsx [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+  name: string;
+  id: string;
+  deleteToken( name: string ): void
+}
+
+export const TokenListItem = ( { name, id, deleteToken }: TokenListItemProps ): JSX.Element => (
+  <ListItem>
+    <>
+      <div>
+        <div className="font-semibold dark:text-gray-100">{name}</div>
```
Comment (MrF3lix) [inline src/editor/components/settings/token-list-item.tsx:14]: I'd like something like this:
image

Use a monospace font, green color and a lock icon before the name. You can use the Heroicon package that is already used in the editor for that.

File: src/editor/pages/api/token/create.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import * as FlooqApi from '../../../helper/flooq-api'
+
+const handler = async ( req: any, res: any ): Promise<void> => {
+  req.body = JSON.parse( req.body )
+  const token = {
+    name: req.body.name,
+    value: req.body.value
+  }
```
Comment (MrF3lix) [inline src/editor/pages/api/token/create.ts:8]: Why do you parse the body, create a new object and then stringify it again? Seems a bit excessive :D

File: src/editor/pages/api/token/delete.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { NextApiRequest, NextApiResponse } from 'next'
+import * as FlooqApi from '../../../helper/flooq-api'
+
+const handler = async ( req: NextApiRequest, res: NextApiResponse ): Promise<void> => {
+  const response = await FlooqApi.deleteRequest( req, `/api/Token/${JSON.parse( req.body ).id}` )
```
Comment (MrF3lix) [inline src/editor/pages/api/token/delete.ts:5]: this is a bit ugly. Instead of passing the id in the body you can use a query param. Then you don't have to parse the body before accessing the id.

File: src/editor/pages/settings.tsx [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+  const [tokens, setTokens] = useState( t )
+  const [globalMessage, setMessage] = useState<Message>( )
+
+  const updateMessage =( message: Message ): void => {
+    setMessage( message )
+    setTimeout( () => setMessage( undefined ), 1500 )
+  }
+
+  const deleteToken = async( id: string ): Promise<void> => {
+    const res = await fetch( `/api/token/delete`, { method: 'DELETE', body: JSON.stringify( { id } ) } )
```
Comment (MrF3lix) [inline src/editor/pages/settings.tsx:29]: use query params instead of body. Usually a request with the DELETE method isn't expected to have a body.

File: src/editor/pages/settings.tsx [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    } else {
+      const res = await fetch( `/api/token/create`, { method: 'POST', body: JSON.stringify( { name, value } ) } )
+      if ( res.ok ) {
+        updateMessage( { text: 'Successfully saved Token.', type: MessageType.Info } )
+        fetch( '/api/token/list' )
+          .then( r => r.json() )
+          .then( setTokens )
+          .catch( () => updateMessage( { text: 'Could not retrieved saved tokens', type: MessageType.Error } ) )
+      } else {
+        updateMessage( { text: 'Error while saving Token.', type: MessageType.Error } )
```
Comment (MrF3lix) [inline src/editor/pages/settings.tsx:50]: Does the API provide an error message? If so we could display it here. But probably it's fine like this.

File: src/editor/pages/settings.tsx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        </div>
+      </main>
+    </>
+  )
+}
+
+
+export const getServerSideProps = async ( context: any ): Promise<{'props': SettingsProps}> => {
+  const tokens = await fetch( `${process.env.BASE_URL}/api/token/list`, { headers: context.req.headers } )
+    .then( res => res.json() )
```
Comment (MrF3lix) [inline src/editor/pages/settings.tsx:77]: we don't use `.then` instead we usually do something like this:

```js
const response = await fetch()
const payload = await response.json()
```

In my opinion this is easier to read and it allows us to check the `response.ok` property to handle errors.

**Review by MrF3lix — CHANGES_REQUESTED** [review]
Good work! Some minor changes required but overall very nice.
image

File: src/editor/components/flow/edit-dialog.tsx (hunk 1 above)
Comment (riesentoaster) [reply src/editor/components/flow/edit-dialog.tsx:22]: Types. I changed it to `useState<string>()` – which you said you preferred.

File: src/editor/pages/api/token/create.ts (hunk 4 above)
Comment (riesentoaster) [reply src/editor/pages/api/token/create.ts:8]: To extract certain attributes from it. How else am I supposed to do this?

File: src/editor/pages/settings.tsx (hunk 8 above)
Comment (riesentoaster) [reply src/editor/pages/settings.tsx:77]: I don't like assigning single-use variables for things like this, it just seems wasteful (even if the waste is negligible). And if you want to check, such a variable is created quickly.

File: src/editor/pages/settings.tsx [hunk 9]
```diff
@@ -38,16 +41,21 @@ export const Settings: NextPage<SettingsProps> = ( { tokens: t }: SettingsProps
...
         updateMessage( { text: 'Successfully saved Token.', type: MessageType.Info } )
-        fetch( '/api/token/list' )
+        return fetch( '/api/token/list' )
           .then( r => r.json() )
           .then( setTokens )
-          .catch( () => updateMessage( { text: 'Could not retrieved saved tokens', type: MessageType.Error } ) )
+          .catch( ( ) => {
+            updateMessage( { text: 'Could not retrieved saved tokens', type: MessageType.Error } )
+            return Promise.reject()
+          } )
```
Comment (MrF3lix) [inline src/editor/pages/settings.tsx:55]: I would use await and try/catch here.

File: src/editor/pages/settings.tsx [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+  return (
+    <>
+      <Head>
+        <title>Flooq | Settings</title>
+      </Head>
+      <PageTitle name="Settings" message={globalMessage} />
+
+      <main>
+        <div className="max-w-7xl mx-auto py-6 sm:px-6 lg:px-8">
+          <List title="Tokens" description="Tokens are used to authenticate Flooq to servers during dataflow execution">
```
Comment (MrF3lix) [inline src/editor/pages/settings.tsx:72]: authentication is just one use case for tokens. I would suggest another title:

> Tokens are stored encrypted and can only be accessed while executing a Data Flow.

**Review by MrF3lix — APPROVED** [review]
